src/datatypes/src/vectors/decimal.rs

```diff
@@ -0,0 +1,465 @@
+
+    fn vector_type_name(&self) -> String {
+        "Decimal128Vector".to_string()
+    }
+
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn len(&self) -> usize {
+        self.array.len()
+    }
+
+    fn to_arrow_array(&self) -> ArrayRef {
+        let data = self.array.to_data();
+        Arc::new(Decimal128Array::from(data))
+    }
+
+    fn to_boxed_arrow_array(&self) -> Box<dyn Array> {
+        let data = self.array.to_data();
+        Box::new(Decimal128Array::from(data))
+    }
+
+    fn validity(&self) -> Validity {
+        vectors::impl_validity_for_vector!(self.array)
+    }
+
+    fn memory_size(&self) -> usize {
+        self.array.get_buffer_memory_size()
+    }
+
+    fn null_count(&self) -> usize {
+        self.array.null_count()
+    }
+
+    fn is_null(&self, row: usize) -> bool {
+        self.array.is_null(row)
+    }
+
+    fn slice(&self, offset: usize, length: usize) -> VectorRef {
+        let data = self.array.to_data().slice(offset, length);
+        Arc::new(Self::from_array_data(data))
+    }
+
+    fn get(&self, index: usize) -> Value {
+        if !self.array.is_valid(index) {
+            return Value::Null;
+        }
+
+        match self.array.data_type() {
+            ArrowDataType::Decimal128(precision, scale) => {
+                // Safety: The index have been checked by `is_valid()`.
+                let value = unsafe { self.array.value_unchecked(index) };
+                Value::Decimal128(Decimal128::new_unchecked(value, *precision, *scale))
+            }
+            _ => Value::Null,
+        }
+    }
+
+    fn get_ref(&self, index: usize) -> ValueRef {
+        if !self.array.is_valid(index) {
+            return ValueRef::Null;
+        }
+
+        match self.array.data_type() {
+            ArrowDataType::Decimal128(precision, scale) => {
+                // Safety: The index have been checked by `is_valid()`.
+                let value = unsafe { self.array.value_unchecked(index) };
+                ValueRef::Decimal128(Decimal128::new_unchecked(value, *precision, *scale))
+            }
+            _ => ValueRef::Null,
+        }
+    }
+}
+
+impl From<Decimal128Array> for Decimal128Vector {
+    fn from(array: Decimal128Array) -> Self {
+        Self { array }
+    }
+}
+
+impl From<Vec<Option<i128>>> for Decimal128Vector {
+    fn from(vec: Vec<Option<i128>>) -> Self {
+        let array = Decimal128Array::from_iter(vec);
+        Self { array }
+    }
+}
+
+impl Serializable for Decimal128Vector {
+    fn serialize_to_json(&self) -> Result<Vec<serde_json::Value>> {
+        self.iter_data()
+            .map(|v| match v {
+                None => Ok(serde_json::Value::Null), // if decimal vector not present, map to NULL
+                Some(vec) => serde_json::to_value(vec),
+            })
+            .collect::<serde_json::Result<_>>()
+            .context(error::SerializeSnafu)
+    }
+}
+
+pub struct Decimal128Iter<'a> {
+    data_type: &'a ArrowDataType,
```

Comment:
Why not store `precision` and `scale` here? You already matched the data_type and got precision in `iter()`.